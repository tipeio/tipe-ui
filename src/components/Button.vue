<template>
  <button
    :class="classObject"
    @click="onClick"
  >
    <slot />
  </button>
</template>

<script>
import vueTypes from 'vue-types'

export default {
  name: 'TipeButton',
  props: {
    size: vueTypes.oneOf(['small', 'medium', 'large']).def('medium'),
    outline: vueTypes.bool.def(false),
    color: vueTypes.oneOf(['purple', 'danger', 'dark-purple']).def('purple')
  },
  computed: {
    classObject: function() {
      const { color, size, outline } = this
      return {
        [size]: true,
        [color]: true,
        outline
      }
    }
  },
  methods: {
    onClick() {
      this.$emit('click')
    }
  }
}
</script>

<style lang="postcss" scoped>
@define-mixin outline $color {
  border: 0.5px solid $color;
  color: $color;
  background-color: #fff;
}

button {
  padding: 10px 20px;
  margin: 10px;
  border-radius: 5px;
  border: none;
  color: #fff;

  &.purple {
    background-image: var(--purple-gradient);
    &.outline {
      background-image: none;
      @mixin outline var(--purple);
    }
  }

  &.dark-purple {
    background-color: var(--purple-dark);
    &.outline {
      @mixin outline var(--purple-dark);
    }
  }

  &.danger {
    background-color: var(--danger);
    &.outline {
      @mixin outline var(--danger);
    }
  }

  &.small {
    font-size: 0.75rem;
  }
  &.medium {
    font-size: 1.25rem;
  }

  &.large {
    font-size: 1.5rem;
  }
}
</style>
