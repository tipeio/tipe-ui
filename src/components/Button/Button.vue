<template>
  <button
    :data-tipe-ui="$options.name"
    :class="classObject"
    @click="onClick"
  >
    <div class="layout">
      <div
        v-if="iconBefore"
        class="icon-container"
      >
        <tipe-icon
          :icon="iconBefore"
          class="icon"
        />
      </div>
      <div
        v-if="$slots.default"
        class="label">
        <slot />
      </div>
      <div
        v-if="iconAfter"
        class="icon-container"
      >
        <tipe-icon
          :icon="iconAfter"
          class="icon"
        />
      </div>
    </div>
  </button>
</template>

<script>
import vueTypes from 'vue-types'
import TipeIcon from '../Icon'

export default {
  name: 'TipeButton',
  components: { TipeIcon },
  props: {
    size: vueTypes.oneOf(['small', 'medium', 'large', 'full']).def('medium'),
    outline: vueTypes.bool.def(false),
    color: vueTypes
      .oneOf(['purple', 'purple-dark', 'gray', 'none'])
      .def('purple'),
    iconBefore: vueTypes.string.def(''),
    iconAfter: vueTypes.string.def('')
  },
  computed: {
    classObject: function() {
      const { color, size, outline } = this
      return {
        [size]: true,
        [color]: true,
        has_icon_before: !!this.iconBefore,
        has_icon_after: !!this.iconAfter,
        outline
      }
    }
  },
  methods: {
    onClick() {
      this.$emit('click')
    }
  }
}
</script>

<style lang="postcss" scoped>
button {
  height: 2.5rem;
  border-radius: 3px;
  margin: 0;
  padding: 0;
  border: none;
  color: #fff;
}

.layout {
  display: inline-grid;
  grid-template-columns: auto auto auto;
  height: 100%;
  color: inherit;
}

.icon-container {
  display: flex;
  justify-content: center;
  align-items: center;
}

button.medium .icon-container,
button.large .icon-container {
  width: 3.125rem;
}

button.small .icon-container {
  width: 1.8125rem;
}

button.medium .icon,
button.large .icon {
  height: 0.8125rem;
  width: 0.8125rem;
}

button.small .icon {
  height: 0.5625rem;
  width: 0.5625rem;
}

.label {
  display: flex;
  justify-content: center;
  align-items: center;
  color: inherit;
  overflow: hidden;
  white-space: nowrap;
  line-height: 0.6875rem;
}

button.medium .label,
button.large .label {
  padding-left: 1.5625rem;
  padding-right: 1.5625rem;
}

button.small .label {
  padding-left: 0.9375rem;
  padding-right: 0.9375rem;
}

button.has_icon_before .label {
  padding-left: 0;
}

button.has_icon_after .label {
  padding-right: 0;
}

/*
Sizes
*/

button.small {
  height: 1.75rem;
}

button.medium {
  height: 2.5rem;
}

button.large {
  height: 3.25rem;
}

button.full {
  width: 100%;
  height: 3.25rem;
}

button.small .label,
button.medium .label {
  font-size: 0.6875rem;
}

button.large .label,
button.full .label {
  font-size: 0.8125rem;
}

/*
Colors
*/

/* fill */
button.purple {
  background-color: var(--purple);
}

button.purple-dark {
  background-color: var(--purple-dark);
}

button.gray {
  background-color: var(--gray-blue-light);
}

button.none {
  background-color: transparent;
}
/* outline */
button.outline {
  background: transparent;
  border-style: solid;
  border-width: 1px;
}

button.outline.purple {
  border-color: var(--purple);
  color: var(--purple);
}

button.outline.purple-dark {
  border-color: var(--purple-dark);
  color: var(--purple-dark);
}

button.outline.gray {
  border-color: var(--gray-blue-light);
  color: var(--text-gray-dark);
}
</style>
