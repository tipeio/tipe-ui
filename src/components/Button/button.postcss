button {
  height: 2.5rem;
  border-radius: 3px;
  margin: 0;
  padding: 0;
  border: none;
  color: #fff;

  &:disabled {
    cursor: not-allowed;
  }
}

.layout {
  display: inline-grid;
  grid-template-columns: auto auto auto;
  height: 100%;
  color: inherit;

  &.dropdown {
    width: 100%;
    & .icon-container {
      justify-content: flex-end;
    }
  }
}

.icon-container {
  display: flex;
  justify-content: center;
  align-items: center;
}

button.medium .icon-container,
button.large .icon-container {
  width: 3.125rem;
}

button.small .icon-container {
  width: 1.8125rem;
}

button.medium .icon,
button.large .icon {
  height: 0.8125rem;
  width: 0.8125rem;
}

button.small .icon {
  height: 0.5625rem;
  width: 0.5625rem;
}

.label {
  display: flex;
  justify-content: center;
  align-items: center;
  color: inherit;
  overflow: hidden;
  white-space: nowrap;
  line-height: 0.6875rem;
}

button.medium .label,
button.large .label {
  padding-left: 1.5625rem;
  padding-right: 1.5625rem;
}

button.small .label {
  padding-left: 0.9375rem;
  padding-right: 0.9375rem;
}

button.has_icon_before .label {
  padding-left: 0;
}

button.has_icon_after .label {
  padding-right: 0;
}

/*
Sizes
*/
button.icon {
  width: 2rem;
  height: 2rem;
}

button.mini {
  width: 6.5625rem;
  height: 1.625rem;
}

button.small {
  width: 6.5625rem;
  height: 1.875rem;
}

button.medium {
  width: 10rem;
  height: 2.5rem;
}

button.large {
  width: 10.5625rem;
  height: 3.125rem;
}

button.full {
  width: 100%;
  height: 3.25rem;
}

button.mini .label,
button.small .label {
  font-size: 0.625rem;
}

button.medium .label {
  font-size: 0.75rem;
}

button.large .label,
button.full .label {
  font-size: 0.9375rem;
}

/*
Colors
*/

/* fill */
button {
  &.primary {
    background-image: var(--primary);
    &:hover {
      background-image: var(--primary-hover);
      box-shadow: var(--box-shawdow);
    }
    &:focus {
      background-image: var(--primary-focus);
    }

  }
  &.success {
    background-color: var(--success);
    color: #fff;
    &:hover {
      background-color: var(--success-hover);
      box-shadow: var(--box-shawdow);
    }
    &:focus {
      background-color: var(--success-focus);
      box-shadow: none;
    }
    &:disabled {
      opacity: 0.4;
      box-shadow: none;
      background-color: var(--disabled);
    }
  }
  &.info {
    background-color: var(--info);
    color: var(--info-text);
    &:hover {
      box-shadow: var(--box-shawdow);
    }
    &:focus {
      background-color: var(--info);
    }
    &:disabled {
      opacity: 0.4;
      box-shadow: none;
      background-color: var(--disabled);
    }
  }
  &.warning {
    background-color: var(--warning);
    color: #fff;
    &:hover {
      background-color: var(--warning-hover);
      box-shadow: var(--box-shawdow);
    }
    &:focus {
      background-color: var(--warning-focus);
    }
    &:disabled {
      opacity: 0.4;
      background-color: var(--disabled);
    }
  }
  &.danger {
    background-color: var(--danger);
    color: #fff;
    &:hover {
      background-color: var(--danger-hover);
      box-shadow: var(--box-shawdow);
    }
    &:focus {
      background-color: var(--danger-focus);
    }
    &:disabled {
      opacity: 0.4;
      background-color: var(--disabled);
    }
  }
}

button.none {
  background-color: transparent;
}

/* outline */
button.outline {
  background: transparent;
  border-style: solid;
  border-width: 1px;

  &.primary {
    border-color: var(--purple);
    color: var(--purple);
    &:hover {
      color: #fff;
      background-image: var(--primary-hover);
      box-shadow: var(--box-shawdow);
    }
    &:focus {
      color: #fff;
      background-image: var(--primary-focus);
    }
  }
  &.success {
    border-color: var(--success);
    color: var(--success);
    &:hover {
      color: #fff;
      background-color: var(--success-hover);
      box-shadow: var(--box-shawdow);
    }
    &:focus {
      color: #fff;
      background-color: var(--success-focus);
      box-shadow: none;
    }
    &:disabled {
      opacity: 0.4;
      box-shadow: none;
      background-color: var(--disabled);
    }
  }
  &.warning {
    border-color: var(--warning);
    color: var(--warning);
    &:hover {
      color: #fff;
      background-color: var(--warning-hover);
      box-shadow: var(--box-shawdow);
    }
    &:focus {
      color: #fff;
      background-color: var(--warning-focus);
    }
    &:disabled {
      opacity: 0.4;
      background-color: var(--disabled);
    }
  }
  &.danger {
    border-color: var(--danger);
    color: var(--danger);
    &:hover {
      color: #fff;
      background-color: var(--danger-hover);
      box-shadow: var(--box-shawdow);
    }
    &:focus {
      color: #fff;
      background-color: var(--danger-focus);
    }
    &:disabled {
      opacity: 0.4;
      background-color: var(--disabled);
    }
  }
  &.info {
    border-color: var(--info);
    color: var(--info);
    &:hover {
      color: #fff;
      background-color: var(--info);
      box-shadow: var(--box-shawdow);
    }
    &:focus {
      color: #fff;
      background-color: var(--info);
    }
    &:disabled {
      opacity: 0.4;
      box-shadow: none;
      background-color: var(--disabled);
    }
  }
}

/*
Icon
*/

button .icon {
  fill: #fff;
}

button.none .icon {
  fill: #000;
}

button.primary.outline .icon {
  fill: var(--purple);
}

button.success.outline .icon {
  fill: var(--success);
}

button.warning.outline .icon {
  fill: var(--warning);
}

button.danger.outline .icon {
  fill: var(--danger);
}

button.outline:hover .icon {
  fill: #fff;
}
