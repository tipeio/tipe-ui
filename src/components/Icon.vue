<template>
  <div
    :class="size"
    :data-tipe-ui="$options.name"
  />
</template>

<script>
import vueTypes from 'vue-types'

const sizes = ['small', 'medium', 'large']

export default {
  name: 'TipeIcon',
  props: {
    size: vueTypes.oneOf(sizes).def('medium')
  }
}
</script>

<style scoped>
[data-tipe-ui='TipeIcon'] {
  background-color: red;
}

[data-tipe-ui='TipeIcon'].small {
  height: 1rem;
  width: 1rem;
}

[data-tipe-ui='TipeIcon'].medium {
  height: 2rem;
  width: 2rem;
}

[data-tipe-ui='TipeIcon'].large {
  height: 3rem;
  width: 3rem;
}
</style>
