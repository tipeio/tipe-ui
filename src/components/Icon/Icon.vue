<template>
  <object
    :data="src"
    type="image/svg+xml">
    <img :src="src" >
  </object>
</template>

<script>
import vueTypes from 'vue-types'

export default {
  name: 'TipeIcon',
  props: {
    icon: vueTypes.string
  },
  computed: {
    src() {
      return require(`../../icons/${this.icon}.svg`)
    }
  }
}
</script>
