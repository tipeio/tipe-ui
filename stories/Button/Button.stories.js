import { storiesOf } from '@storybook/vue'
import { action } from '@storybook/addon-actions'

import TipeButton from '@/components/Button'

import '@/styles/globals.css'

storiesOf('Button', module)
  .add('color: primary', () => ({
    components: { TipeButton },
    template: '<tipe-button color="primary">Fill Button</tipe-button>',
    methods: { action: action('clicked') }
  }))
  .add('color: info', () => ({
    components: { TipeButton },
    template: '<tipe-button color="info">Fill Button</tipe-button>',
    methods: { action: action('clicked') }
  }))
  .add('color: success', () => ({
    components: { TipeButton },
    template: '<tipe-button color="success">Fill Button</tipe-button>',
    methods: { action: action('clicked') }
  }))
  .add('color: warning', () => ({
    components: { TipeButton },
    template: '<tipe-button color="warning">Fill Button</tipe-button>',
    methods: { action: action('clicked') }
  }))
  .add('color: danger', () => ({
    components: { TipeButton },
    template: '<tipe-button color="danger">Fill Button</tipe-button>',
    methods: { action: action('clicked') }
  }))
  .add('outline', () => ({
    components: { TipeButton },
    template: '<tipe-button outline>Fill Button</tipe-button>',
    methods: { action: action('clicked') }
  }))
  .add('disabled', () => ({
    components: { TipeButton },
    template: '<tipe-button :disabled="true">Fill Button</tipe-button>',
    methods: { action: action('clicked') }
  }))
  .add('size: mini', () => ({
    components: { TipeButton },
    template: '<tipe-button size="mini">Mini Button</tipe-button>',
    methods: { action: action('clicked') }
  }))
  .add('size: small', () => ({
    components: { TipeButton },
    template: '<tipe-button size="small">Small Button</tipe-button>',
    methods: { action: action('clicked') }
  }))
  .add('size: medium', () => ({
    components: { TipeButton },
    template: '<tipe-button size="medium">Medium Button</tipe-button>',
    methods: { action: action('clicked') }
  }))
  .add('size: large', () => ({
    components: { TipeButton },
    template: '<tipe-button size="large">Large Button</tipe-button>',
    methods: { action: action('clicked') }
  }))
  .add('icon before label', () => ({
    components: { TipeButton },
    template: '<tipe-button iconBefore="plus">Icon Button</tipe-button>',
    methods: { action: action('clicked') }
  }))
  .add('icon after label', () => ({
    components: { TipeButton },
    template: '<tipe-button iconAfter="plus">Icon Button</tipe-button>',
    methods: { action: action('clicked') }
  }))
  .add('icon before and after label', () => ({
    components: { TipeButton },
    template:
      '<tipe-button iconBefore="plus" iconAfter="plus">Icon Button</tipe-button>',
    methods: { action: action('clicked') }
  }))
  .add('icon only', () => ({
    components: { TipeButton },
    template: '<tipe-button color="primary" iconBefore="plus"></tipe-button>',
    methods: { action: action('clicked') }
  }))
  .add('icon only no backgound', () => ({
    components: { TipeButton },
    template: '<tipe-button color="none" iconBefore="plus"></tipe-button>',
    methods: { action: action('clicked') }
  }))
