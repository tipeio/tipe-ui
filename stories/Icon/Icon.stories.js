import { storiesOf } from '@storybook/vue'

import TipeIcon from '../../src/components/Icon'

const style = () => ({
  height: '1rem',
  width: '1rem'
})

const icons = [
  'check',
  'document',
  'folder',
  'key',
  'number',
  'share',
  'switch',
  'user',
  'color',
  'edit',
  'gear',
  'link',
  'pin',
  'sign-out',
  'text',
  'view',
  'copy',
  'flow',
  'grid',
  'list',
  'search',
  'star-active',
  'trash',
  'wrench',
  'credit-card',
  'folder-open',
  'group',
  'markdown',
  'media',
  'send',
  'star',
  'user-circle',
  'options',
  'plus',
  'cross',
  'arrow-down',
  'arrow-up',
  'arrow-right',
  'arrow-left'
]

storiesOf('Icon', module).add('icons', () => ({
  components: { TipeIcon },
  computed: { style },
  template:
    '<div>' +
    icons.reduce((template, icon) => {
      return (
        template +
        `<tipe-icon :style="style" icon="${icon}" color="rgb(255,0,0)" />`
      )
    }, '') +
    '</div>'
}))
