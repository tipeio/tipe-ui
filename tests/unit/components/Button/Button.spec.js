import { shallowMount } from '@vue/test-utils'
import { createRenderer } from 'vue-server-renderer'
import tipeTestUtils from '~/tests/TipeTestUtils'
import Button from '@/components/Button'
import Icon from '@/components/Icon'

describe('Button.vue', () => {
  describe('<template>', () => {
    it('matches previous snapshot', async () => {
      const wrapper = shallowMount(Button)
      await expect(
        createRenderer().renderToString(wrapper.vm)
      ).toMatchSnapshot()
    })

    it('has correct data-tipe-iu attibute', () => {
      const wrapper = shallowMount(Button)
      expect(wrapper.attributes()['data-tipe-ui']).toBe('TipeButton')
    })
  })

  describe(':props', () => {
    tipeTestUtils.test.input.size(Button)
    tipeTestUtils.test.input.status(Button)
    tipeTestUtils.test.input.waiting(Button)
    tipeTestUtils.test.input.disabled(Button)

    it(':color(primary) - has primary class', () => {
      const wrapper = shallowMount(Button, {
        propsData: { color: 'primary' }
      })

      expect(wrapper.classes()).toContain('primary')
    })

    it(':color(none) - has none class', () => {
      const wrapper = shallowMount(Button, {
        propsData: { color: 'none' }
      })

      expect(wrapper.classes()).toContain('none')
    })

    it(':iconBefore - renders icon before', () => {
      const propsData = { iconBefore: 'plus' }
      const wrapper = shallowMount(Button, { propsData })

      expect(wrapper.contains('.icon-container')).toEqual(true)
      expect(wrapper.classes()).toContain('has_icon_before')
      expect(wrapper.contains(Icon)).toEqual(true)
      expect(wrapper.find(Icon).props().icon).toEqual(propsData.iconBefore)
    })

    it(':iconAfter - renders icon after', () => {
      const propsData = { iconAfter: 'plus' }
      const wrapper = shallowMount(Button, { propsData })

      expect(wrapper.contains('.icon-container')).toEqual(true)
      expect(wrapper.classes()).toContain('has_icon_after')
      expect(wrapper.contains(Icon)).toEqual(true)
      expect(wrapper.find(Icon).props().icon).toEqual(propsData.iconAfter)
    })
  })

  describe('@events', () => {
    it('@click - should emit', () => {
      const wrapper = shallowMount(Button)

      wrapper.find('button').trigger('click')
      expect(wrapper.emitted().click).toBeTruthy()
    })
  })
})
