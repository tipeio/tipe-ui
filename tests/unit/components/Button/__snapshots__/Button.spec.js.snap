// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button.vue <template> matches previous snapshot 1`] = `Promise {}`;
